### PR TITLE
Accept supplementary Unicode characters in identifiers

### LINKED ISSUE #1406: Scala should support Unicode supplementary characters
Currently the parser only understands identifiers drawn from the basic multilingual plane.  

See https://www.oracle.com/technical-resources/articles/javase/supplementary.html for more information on how we would support this on the Java target.

### PR DESCRIPTION
**Pull request by som-snytt**
Previously the lexer only accepted identifiers drawn from the basic multilingual plane. 

Fixes scala/bug#1406

### CODE REVIEW

File: build.sbt
Comment (som-snytt) [inline build.sbt]: "-Wconf:cat=lint-nullary-unit&site=.*Test:s"

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:703]: suspicious empty lit case

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (martijnhoekstra) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: To keep accepting `val high = '\ud801'` (and also the more questionable literal version of that sequence), maybe something like this?

```suggestion
                  if (cbuf.length == 2)
```

Comment (sjrd) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: Indeed, it's important to allow `'\uhhhh'` escapes for surrogate chars. Also to allow unpaired surrogate chars in `\uhhhh` form within string literals. Both are important to write tests of string-related operations that need to deal with this stuff.

Comment (martijnhoekstra) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: Invalid unicode string literals are unaffected by this PR and are still accepted, it's just the unpaired high surrogate escape char literals.

If this goes through `\Uhhhhhh` escapes in addition to `\uhhhh` escapes would also be neat.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: Thanks, that crossed my mind but I wasn't sure of the use case or test, see the new test. That relaxes both the `\u` example and having an orphan surrogate in input. Note that under the old unicode regime, escape was processed early and the two cases were not distinguishable. Also noticed backquoted idents are literal chars that fall under the purview.

**Comment by martijnhoekstra** [comment]
If I read this correctly, the biggest change is in plain identifiers, which now allow non-BMP characters if allowed by the identifier grammar. Strings literals, interpolations, back quoted identifiers, char literals and other literals are all unchanged: even where the spec didn't allow supplementary characters, the implementation did.

Changed:

* Plain identifiers:
  * non-BMP characters are allowed (if allowed by the identifier grammar) (new in this PR)
  * Invalid codepoints or code sequences are not allowed in any way, shape or form (unchanged)
* ~Identifiers in string interpolations~
  * but maybe that is wrong! ✅ 
* Chars literals:
  * non-BMP characters are consumed but not allowed (they don't fit)

Unchanged:

* Strings:
  * non-BMP characters are accepted (no change)
  * Invalid codepoints are allowed (no change)
  * Invalid code sequences (unpaired/mismatched surrogates) are allowed if they're written as Unicode escapes (no change)
  * invalid code sequences  (unpaired/mismatched surrogates) are allowed if they somehow made it to the scanner (no change). Unsure how/if that's possible in different source implementations.
 
* Interpolations: same as strings, but without interpreting escapes (which is up to the interpolator) (no change)

* Chars literals:
  * Invalid code sequences (unpaired surrogates, other non-characters like U+FEFF) are allowed as unicode escapes
  * The same invalid code sequences would be allowed if they make it to the scanner, which may be impossible

* Other literals:
  * moot, it's not like, say, integer literals would accept any non-BMP characters anyway.

How are back-quoted identifiers handled here? What do they accept and does that break on any backend? -- edit: I checked this, they are properly encoded.

Implementation wise it's really nice. 

This change is not yet reflected in the spec. Maybe spec work can be done separately from this PR? That's up to the lightbend team to decide I think. Same for scalameta and scala 3 parity.

**Comment by lrytz** [comment]
Thanks for the overview @martijnhoekstra

From what I can tell this LGTM, it aligns with Java. Agree that it's important to align Scala 3, scalameta, ping IntelliJ.

> This change is not yet reflected in the spec.

Are you sure? Isn't this PR fixing the implementation to match the spec?

**Comment by SethTisue** [comment]
@som-snytt are you interested in forward-porting to Scala 3?

**Comment by martijnhoekstra** [comment]
> Are you sure? Isn't this PR fixing the implementation to match the spec?

Yes, on the very first line :)

> Scala programs are written using the Unicode Basic Multilingual Plane (BMP) character set; Unicode supplementary characters are not presently supported.

Which could be outright deleted or maybe replaced with "Scala programs are written using the Unicode character set"

scalac has been more lenient than the spec mandates in it's input for as long as I've known and accepted non-BMP characters in strings for example.

```
class AlwaysHasBeen {
  final val `𐐀` = "𐐀 has always been allowed"
  final val always: "𐐀 has always been allowed" = "\ud801\udc00 has always been allowed"
}
```

was already accepted (and still is).

**Comment by martijnhoekstra** [comment]
scala also accepts (and has accepted) surrogate characters in UTF-8. UTF-8 has no surrogate characters, and these code sequences are not "real" Unicode. So if you're a stickler for both the scala spec and the Unicode spec, you can argue that the spec is incorrect because of that, and in many places where it says it accepts an Unicode character, it actually accepts a codepoint that doesn't encode an unicode Character.

The problem is there isn't anyone you can argue that *to*, because by the time you're halfway to the point you're trying to make, their eyes have long glazed over, which I suppose is completely justified.

**Comment by som-snytt** [comment]
@martijnhoekstra That's right, anything going through `getLitChar` just takes a char as data. I didn't even need that residual call to `isSupplementary` there, so I will remove it and squash those two commits after your previous review. (To recap: my first pass disallowed "orphan surrogates" in literals.)

I'll also add test for string interpolation with supplementary char in constant part; if parser handles the interpolated identifier, then it just vacuums up the constants.

I'll update the spec.

@SethTisue yes, I will forward port, now that Martijn has explained to me what I did. Also, I just right-clicked to add Martijn to the dictionary and get rid of the red squiggle.

**Comment by som-snytt** [comment]
To the last mindbending point, all this is an artifact of JVM support. If we were ingesting codepoints, the only check would be that we read a valid codepoint, as we must do when assembling a supplementary character from surrogates.

I was going to tweet that day that I tried to think of "high surrogate" and came up with "high suffragette".

On "non-Unicode" or "broken" characters, I'm still wondering if it's worth accepting non-chars in unicode escape form only. I will follow up later if that is deemed useful.

**Comment by martijnhoekstra** [comment]
Accepting broken characters in source form is pretty hard: by the time you go from a source encoding to an Array[Char], the encoder should have already rejected the bad sequences.

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (martijnhoekstra) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: Nit: can we be more specific about the illegality in the message?

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (martijnhoekstra) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:1105]: Nice!

**Review by martijnhoekstra — APPROVED** [review]

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: Not a nit. I had an update and squash earlier but I didn't like it so I'll try again. Thanks for review.

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:1041]: Sorry, guy, but this is totally wrong. `s"$id"` requires alpha to start and alnum plus underscore for rest.

**Comment by som-snytt** [comment]
By broken char, I meant run/t1406b, where we read a high surrogate only (not a unicode escape). As you point out, it's just a non-character, so is it more special than other non-character characters?

**Comment by som-snytt** [comment]
Touched up the spec. The first sentence really needs to grab the reader.

Noticed that it would be nice to support varid starting with supplementary char, as that is defined in terms of letter and lower. As noted in the review, supplementary chars are encoded, and it's probably excessive work to change that, though not necessarily undesirable. The minimal change is for isVariableName to decode if necessary.

**Comment by som-snytt** [comment]
Retouched up the spec.

File: spec/01-lexical-syntax.md
Comment (sjrd) [inline spec/01-lexical-syntax.md]: Consider writing "Unicode code points" instead of "Unicode characters". Unicode has, on purpose, no notion of "character" (the closest that comes to the human notion of character is in fact a Unicode Extended Cluster).

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:1041]: Actually, why did he say that? it's perfectly fine to test isUnicodeIdentifier. Maybe he was thinking how they rejected accepting arbitrary ids, operator ids and backticks.

File: spec/01-lexical-syntax.md
Comment (som-snytt) [reply spec/01-lexical-syntax.md]: Thanks, it crossed my mind. I'll reconsider. "Scala code is Unicode code points!" It's points of code!

**Comment by som-snytt** [comment]
"Supplimentary" is when you try to be nice and supportive, but also suave and supple.

Edit: suppliments are the complimentary supplements they give out on race day.

File: spec/01-lexical-syntax.md
Comment (lrytz) [inline spec/01-lexical-syntax.md]: > certain character classes

Does this refer to the list of classes below?

Does the first half of this paragraph add something to the spec, or only clarify?

Comment (som-snytt) [reply spec/01-lexical-syntax.md]: I added it as clarification in response to sjrd's comment about "character". Unicode does say "character", in our sense as opposed to "what people take as a single character", but OTOH it's inherently complex or confusing. I'm happy to re-edit as many times as necessary; I'm not wedded to any wording.

IMHO, "Scala source text consists of Unicode code points" makes me wonder if they're saying something subtle. But my perception may be warped by writing code. Maybe "Scala supports Unicode source text."

The clarification of "What do you mean, you support Unicode?" is that a math symbol `+` that is not the ascii is taken as an operator, though maybe not the unary one you're thinking of, and that any code point that is a letter can start an identifier, not just BMP. The intention also is to be vague enough not to say something wrong or too much.

Comment (martijnhoekstra) [reply spec/01-lexical-syntax.md]: Another alternative to consider: don't mention that scala supports or is supported by unicode (non)characters and code points, just mention that the spec is defines the scala grammar in terms of unicode:

> The scala lexical grammar is expressed in Unicode code points and Unicode character properties.

Comment (lrytz) [reply spec/01-lexical-syntax.md]: For the spec I think it's fine to just say "Unicode code points", it's precise and googleable.

Comment (SethTisue) [reply spec/01-lexical-syntax.md]: agree, but I'd also be fine with "Scala source code consists of Unicode text"

**Comment by SethTisue** [comment]
I suggest we hold off on merging this until the Scala 3 forward-port has landed. (I don't like to put the Scala 3 folks on the hook for something without their consent, and also the forward-port might produce additional useful review feedback.)

File: spec/01-lexical-syntax.md
Comment (som-snytt) [reply spec/01-lexical-syntax.md]: I'll go with Seth's minimization since he holds the merge button.

**Comment by som-snytt** [comment]
OK, I'll go ahead and attempt to forward port. I think the metaphor for "on the hook" here is "catch and release". You put the immature feature back into the ecosystem until it is big enough for frying.

File: spec/01-lexical-syntax.md
Comment (som-snytt) [reply spec/01-lexical-syntax.md]: As I typed it in, I just noticed how Seth cleverly moved "code" to the left and "text" to the right!

Comment (som-snytt) [reply spec/01-lexical-syntax.md]: Also Scala turtles consist of Uniturtle turtles. It's turtles all the way...

**Comment by som-snytt** [comment]
Had to move a dotty test out of the way, born of that necessity was
```
tests/neg/sorrygates.scala
```

**Comment by SethTisue** [comment]
seems clear the Scala 3 forward port will land, so...

**Comment by SethTisue** [comment]
Thank you @som-snytt! Let a thousand monstrous, bizarre identifier names bloom! 👹

**Comment by unkarjedy** [comment]
IntelliJ issue https://youtrack.jetbrains.com/issue/SCL-19655

**Comment by unkarjedy** [comment]
https://www.scala-lang.org/files/archive/spec/2.13/13-syntax-summary.html
```
opchar           ::=  ‘!’ | ‘#’ | ‘%’ | ‘&’ | ‘*’ | ‘+’ | ‘-’ | ‘/’ | ‘:’ |
                      ‘<’ | ‘=’ | ‘>’ | ‘?’ | ‘@’ | ‘\’ | ‘^’ | ‘|’ | ‘~’
                      and any character in Unicode categories Sm or So
```

Supplementary unicode characters `\u1F300..\u1F5FF` are also included by  Unicode So category.
See `java.lang.Character.UnicodeBlock#MISCELLANEOUS_SYMBOLS_AND_PICTOGRAPHS`
But they are not parsed in `2.13.7`
```scala
class Suppl {
  def ☀ = 42
  def 𐐀 = 42
  def 🌂 = 42
  def 🌀 = 42
}

**Comment by som-snytt** [comment]
adriaanm or retronym would have said but now you'll have to notify the IDEs, which used to mean the eclipse plugin first of all. tpolecat already complained that this repo does not sync with its tooling consumers on any level. This was my fault, as I am lazy and not on a first name basis with any maintainers.

Edit: "seems clear the Scala 3 forward port will land, so..." or maybe not?

**Comment by martijnhoekstra** [comment]
Everyone wants to be on a first name basis with you, Som. I knew a guy named Chip, he had a cat named Munk. Anyway, which first names should we have pinged?

**Comment by SethTisue** [comment]
followup PR landed in 2.13.9: https://github.com/scala/scala/pull/9805

**Pull request merged**
